File: backend/internal/services/summary.go
package services

import (
	"context"
	"encoding/json"
	"net/http"
	"sync"
	"time"

	"github.com/petrovito/china-trip-bookings/backend/internal/models"
)

// ── Expense type set ──────────────────────────────────────────────────────────

// AllExpenseTypes lists booking types that count as expenses (in display order).
var AllExpenseTypes = []string{
	"flight", "hotel", "train", "ticket", "food", "city_transport", "shopping",
}

var expenseTypeSet = map[string]bool{
	"flight": true, "hotel": true, "train": true, "ticket": true,
	"food": true, "city_transport": true, "shopping": true,
}

// ── Exchange rate cache ───────────────────────────────────────────────────────

var (
	rateMu        sync.Mutex
	cachedRates   map[string]float64
	cachedRatesAt string
	ratesFetchedAt time.Time
)

// FetchRates returns DKK-base exchange rates, caching for 30 minutes.
// Returns stale data on error if available.
func FetchRates() (rates map[string]float64, ratesAt string, err error) {
	rateMu.Lock()
	defer rateMu.Unlock()

	if cachedRates != nil && time.Since(ratesFetchedAt) < 30*time.Minute {
		return cachedRates, cachedRatesAt, nil
	}

	client := &http.Client{Timeout: 6 * time.Second}
	resp, fetchErr := client.Get("https://open.er-api.com/v6/latest/DKK")
	if fetchErr != nil {
		return cachedRates, cachedRatesAt, fetchErr
	}
	defer resp.Body.Close()

	var result struct {
		Result         string             `json:"result"`
		TimeLastUpdate string             `json:"time_last_update_utc"`
		Rates          map[string]float64 `json:"rates"`
	}
	if decErr := json.NewDecoder(resp.Body).Decode(&result); decErr != nil {
		return cachedRates, cachedRatesAt, decErr
	}
	if result.Result == "success" && len(result.Rates) > 0 {
		cachedRates = result.Rates
		cachedRatesAt = result.TimeLastUpdate
		ratesFetchedAt = time.Now()
	}
	return cachedRates, cachedRatesAt, nil
}

// ── DKK conversion helpers ────────────────────────────────────────────────────

// toDKK converts an amount to DKK using DKK-base rates.
// Returns (0, false) if the currency rate is unavailable.
func toDKK(amount float64, currency string, rates map[string]float64) (float64, bool) {
	if rates == nil {
		return 0, false
	}
	if currency == "" || currency == "DKK" {
		return amount, true
	}
	rate, ok := rates[currency]
	if !ok || rate == 0 {
		return 0, false
	}
	return amount / rate, true
}

// ── Section builders ──────────────────────────────────────────────────────────

func filterByTravelers(bks []models.Booking, who string) []models.Booking {
	out := []models.Booking{}
	for _, b := range bks {
		switch who {
		case "": // collective = travelers is "both" or empty
			if b.Travelers == "both" || b.Travelers == "" {
				out = append(out, b)
			}
		default:
			if b.Travelers == who {
				out = append(out, b)
			}
		}
	}
	return out
}

func buildExpenseSection(bks []models.Booking, rates map[string]float64) models.ExpenseSection {
	byType := map[string]models.TypeExpenses{}
	section := models.ExpenseSection{ByType: byType}

	for _, t := range AllExpenseTypes {
		items := []models.Booking{}
		for _, b := range bks {
			if b.Type == t {
				items = append(items, b)
			}
		}
		if len(items) == 0 {
			continue
		}
		te := models.TypeExpenses{Items: items}
		for _, b := range items {
			if b.Price == nil {
				continue
			}
			dkk, ok := toDKK(*b.Price, b.Currency, rates)
			if !ok {
				te.Approx = true
				section.Approx = true
			} else {
				te.TotalDKK += dkk
				section.TotalDKK += dkk
			}
		}
		byType[t] = te
	}
	return section
}

func buildSettlement(bks []models.Booking, rates map[string]float64) models.SettlementData {
	sd := models.SettlementData{ByType: map[string]models.TypeSettlement{}}

	for _, t := range AllExpenseTypes {
		items := []models.Booking{}
		for _, b := range bks {
			if b.Type == t {
				items = append(items, b)
			}
		}
		if len(items) == 0 {
			continue
		}
		ts := models.TypeSettlement{Items: items}
		for _, b := range items {
			if b.Price == nil || b.PaidBy == nil {
				continue
			}
			dkk, ok := toDKK(*b.Price, b.Currency, rates)
			if !ok {
				ts.Approx = true
				sd.Approx = true
				continue
			}
			// positive contrib = peter owes friend; negative = friend owes peter
			var contrib float64
			if *b.PaidBy == "friend" {
				contrib = dkk / 2
			} else {
				contrib = -dkk / 2
			}
			if b.Settled {
				ts.SettledDKK += contrib
				sd.NetSettledDKK += contrib
			} else {
				ts.OutstandingDKK += contrib
				sd.NetOutstandingDKK += contrib
			}
		}
		ts.PeterOwes = ts.OutstandingDKK > 0.5
		ts.FriendOwes = ts.OutstandingDKK < -0.5
		sd.ByType[t] = ts
	}

	sd.PeterOwes = sd.NetOutstandingDKK > 0.5
	sd.FriendOwes = sd.NetOutstandingDKK < -0.5
	return sd
}

// ── Main entry point ──────────────────────────────────────────────────────────

// ComputeFullSummary builds the complete summary including DKK conversion.
func ComputeFullSummary(ctx context.Context) (*models.FullSummaryResponse, error) {
	bookings, err := FetchAllBookings(ctx)
	if err != nil {
		return nil, err
	}

	rates, ratesAt, _ := FetchRates() // ignore fetch error; nil rates → approx: true

	var ratesAtPtr *string
	if ratesAt != "" {
		s := ratesAt
		ratesAtPtr = &s
	}

	// Filter to expense bookings with a price
	var expBks []models.Booking
	for _, b := range bookings {
		if expenseTypeSet[b.Type] && b.Price != nil {
			expBks = append(expBks, b)
		}
	}

	collective := filterByTravelers(expBks, "")
	peterBks   := filterByTravelers(expBks, "peter")
	friendBks  := filterByTravelers(expBks, "friend")

	colSection := buildExpenseSection(collective, rates)
	petSection := buildExpenseSection(peterBks, rates)
	friSection := buildExpenseSection(friendBks, rates)

	// Per-person totals: personal + half of shared
	petTotal  := colSection.TotalDKK/2 + petSection.TotalDKK
	friTotal  := colSection.TotalDKK/2 + friSection.TotalDKK
	petApprox := colSection.Approx || petSection.Approx
	friApprox := colSection.Approx || friSection.Approx

	// Settlement from paid shared bookings
	var sharedPaid []models.Booking
	for _, b := range collective {
		if b.PaidBy != nil {
			sharedPaid = append(sharedPaid, b)
		}
	}
	settlement := buildSettlement(sharedPaid, rates)

	// Per-segment expense totals
	bySegment := map[string]models.SegmentExpenses{}
	for _, b := range expBks {
		if b.SegmentID == nil {
			continue
		}
		k := b.SegmentID.String()
		se := bySegment[k]
		if se.ByType == nil {
			se.ByType = map[string]models.DKKAmount{}
		}
		if b.Price == nil {
			bySegment[k] = se
			continue
		}
		dkk, ok := toDKK(*b.Price, b.Currency, rates)
		cur := se.ByType[b.Type]
		if !ok {
			cur.Approx = true
			se.Approx = true
		} else {
			cur.TotalDKK += dkk
			se.TotalDKK += dkk
		}
		se.ByType[b.Type] = cur
		bySegment[k] = se
	}

	return &models.FullSummaryResponse{
		Rates:   rates,
		RatesAt: ratesAtPtr,
		Collective: colSection,
		Peter:      petSection,
		Friend:     friSection,
		PerPerson: models.PerPersonTotals{
			Peter:  models.DKKAmount{TotalDKK: petTotal, Approx: petApprox},
			Friend: models.DKKAmount{TotalDKK: friTotal, Approx: friApprox},
		},
		Settlement: settlement,
		BySegment:  bySegment,
	}, nil
}

File: flutter/lib/models/summary.dart
import 'booking.dart';

class SummaryData {
  final Map<String, double> rates;
  final String? ratesAt;
  final ExpenseSection collective;
  final ExpenseSection peter;
  final ExpenseSection friend;
  final PerPersonTotals perPerson;
  final SettlementData settlement;
  final Map<String, SegmentExpenses> bySegment;

  const SummaryData({
    this.rates = const {},
    this.ratesAt,
    required this.collective,
    required this.peter,
    required this.friend,
    required this.perPerson,
    required this.settlement,
    this.bySegment = const {},
  });

  factory SummaryData.fromJson(Map<String, dynamic> j) => SummaryData(
    rates:      (j['rates'] as Map<String, dynamic>? ?? {})
        .map((k, v) => MapEntry(k, (v as num).toDouble())),
    ratesAt:    j['rates_at']?.toString(),
    collective: ExpenseSection.fromJson(j['collective'] as Map<String, dynamic>? ?? {}),
    peter:      ExpenseSection.fromJson(j['peter'] as Map<String, dynamic>? ?? {}),
    friend:     ExpenseSection.fromJson(j['friend'] as Map<String, dynamic>? ?? {}),
    perPerson:  PerPersonTotals.fromJson(j['per_person'] as Map<String, dynamic>? ?? {}),
    settlement: SettlementData.fromJson(j['settlement'] as Map<String, dynamic>? ?? {}),
    bySegment:  (j['by_segment'] as Map<String, dynamic>? ?? {})
        .map((k, v) => MapEntry(k, SegmentExpenses.fromJson(v as Map<String, dynamic>))),
  );
}

class ExpenseSection {
  final Map<String, TypeExpenses> byType;
  final double totalDKK;
  final bool approx;

  const ExpenseSection({this.byType = const {}, this.totalDKK = 0, this.approx = false});

  factory ExpenseSection.fromJson(Map<String, dynamic> j) => ExpenseSection(
    byType:   (j['by_type'] as Map<String, dynamic>? ?? {})
        .map((k, v) => MapEntry(k, TypeExpenses.fromJson(v as Map<String, dynamic>))),
    totalDKK: (j['total_dkk'] as num?)?.toDouble() ?? 0,
    approx:   j['approx'] as bool? ?? false,
  );
}

class TypeExpenses {
  final List<Booking> items;
  final double totalDKK;
  final bool approx;

  const TypeExpenses({this.items = const [], this.totalDKK = 0, this.approx = false});

  factory TypeExpenses.fromJson(Map<String, dynamic> j) => TypeExpenses(
    items:    (j['items'] as List? ?? [])
        .map((b) => Booking.fromJson(b as Map<String, dynamic>))
        .toList(),
    totalDKK: (j['total_dkk'] as num?)?.toDouble() ?? 0,
    approx:   j['approx'] as bool? ?? false,
  );
}

class PerPersonTotals {
  final DKKAmount peter;
  final DKKAmount friend;

  const PerPersonTotals({required this.peter, required this.friend});

  factory PerPersonTotals.fromJson(Map<String, dynamic> j) => PerPersonTotals(
    peter:  DKKAmount.fromJson(j['peter'] as Map<String, dynamic>? ?? {}),
    friend: DKKAmount.fromJson(j['friend'] as Map<String, dynamic>? ?? {}),
  );
}

class DKKAmount {
  final double totalDKK;
  final bool approx;

  const DKKAmount({this.totalDKK = 0, this.approx = false});

  factory DKKAmount.fromJson(Map<String, dynamic> j) => DKKAmount(
    totalDKK: (j['total_dkk'] as num?)?.toDouble() ?? 0,
    approx:   j['approx'] as bool? ?? false,
  );
}

class SettlementData {
  final double netOutstandingDKK;
  final double netSettledDKK;
  final bool peterOwes;
  final bool friendOwes;
  final bool approx;
  final Map<String, TypeSettlement> byType;

  const SettlementData({
    this.netOutstandingDKK = 0,
    this.netSettledDKK = 0,
    this.peterOwes = false,
    this.friendOwes = false,
    this.approx = false,
    this.byType = const {},
  });

  factory SettlementData.fromJson(Map<String, dynamic> j) => SettlementData(
    netOutstandingDKK: (j['net_outstanding_dkk'] as num?)?.toDouble() ?? 0,
    netSettledDKK:     (j['net_settled_dkk'] as num?)?.toDouble() ?? 0,
    peterOwes:         j['peter_owes'] as bool? ?? false,
    friendOwes:        j['friend_owes'] as bool? ?? false,
    approx:            j['approx'] as bool? ?? false,
    byType:            (j['by_type'] as Map<String, dynamic>? ?? {})
        .map((k, v) => MapEntry(k, TypeSettlement.fromJson(v as Map<String, dynamic>))),
  );
}

class TypeSettlement {
  final List<Booking> items;
  final double outstandingDKK;
  final double settledDKK;
  final bool peterOwes;
  final bool friendOwes;
  final bool approx;

  const TypeSettlement({
    this.items = const [],
    this.outstandingDKK = 0,
    this.settledDKK = 0,
    this.peterOwes = false,
    this.friendOwes = false,
    this.approx = false,
  });

  factory TypeSettlement.fromJson(Map<String, dynamic> j) => TypeSettlement(
    items:          (j['items'] as List? ?? [])
        .map((b) => Booking.fromJson(b as Map<String, dynamic>))
        .toList(),
    outstandingDKK: (j['outstanding_dkk'] as num?)?.toDouble() ?? 0,
    settledDKK:     (j['settled_dkk'] as num?)?.toDouble() ?? 0,
    peterOwes:      j['peter_owes'] as bool? ?? false,
    friendOwes:     j['friend_owes'] as bool? ?? false,
    approx:         j['approx'] as bool? ?? false,
  );
}

class SegmentExpenses {
  final Map<String, DKKAmount> byType;
  final double totalDKK;
  final bool approx;

  const SegmentExpenses({this.byType = const {}, this.totalDKK = 0, this.approx = false});

  factory SegmentExpenses.fromJson(Map<String, dynamic> j) => SegmentExpenses(
    byType:   (j['by_type'] as Map<String, dynamic>? ?? {})
        .map((k, v) => MapEntry(k, DKKAmount.fromJson(v as Map<String, dynamic>))),
    totalDKK: (j['total_dkk'] as num?)?.toDouble() ?? 0,
    approx:   j['approx'] as bool? ?? false,
  );
}
